packages/@aws-cdk/aws-lambda/lib/function.ts

```diff
@@ -1030,31 +1041,46 @@ Environment variables can be marked for removal when used in Lambda@Edge by sett
     };
   }
 
+  private isQueue(deadLetterQueue: sqs.IQueue | sns.ITopic): deadLetterQueue is sqs.IQueue {
+    return (<sqs.IQueue>deadLetterQueue).queueArn !== undefined;
+  }
+
   private buildDeadLetterQueue(props: FunctionProps) {
     if (props.deadLetterQueue && props.deadLetterQueueEnabled === false) {
       throw Error('deadLetterQueue defined but deadLetterQueueEnabled explicitly set to false');
     }
 
-    if (!props.deadLetterQueue && !props.deadLetterQueueEnabled) {
+    let deadLetterQueue: sqs.IQueue | sns.ITopic;
+
+    if (!props.deadLetterQueue && !props.deadLetterQueueEnabled && !props.deadLetterTopic) {
       return undefined;
+    } else if (props.deadLetterTopic && props.deadLetterQueue) {
+      throw new Error('deadLetterQueue and deadLetterTopic cannot be specified together at the same time');
+    } else if (props.deadLetterTopic && (props.deadLetterQueueEnabled === false || props.deadLetterQueueEnabled === true)) {
+      throw new Error('deadLetterQueueEnabled and deadLetterTopic cannot be specified together at the same time');
+    } else if (props.deadLetterTopic && !props.deadLetterQueueEnabled) {
```

Comment:
i dont love these chained ifs. I would prefer we separate them out into two ifs that are checking for errors and `undefined`, and then an if that determines if we create a queue or topic:

```ts
if (!props.dlq && !props.dlqe && !props.dlt) {
  return undefined;
}
if (props.dlt && props.dlq || props.dlqe !== undefined) { throw new Error(); }

// props.dlq or props.dlt must be enabled here
if (props.dlt) { ... } else { ... }
```

Reply:
Alright. I don't actually mind chaining if-s but you're right that amount of information to digest in conditions + chaining makes thing increadibly blurred.